File: backupGiocoPy/pedineNoAnimazione/GiocoPython/crazyGoose.py
from Giocatore_modul import *
from Percorso_modul import *
from Casella_modul import *
from menu import *
from globalFunction import *

import pygame
import random
import time
import threading

INFO_PL1 = "Tu (PL1)"
INFO_COM = "Computer (COM)"
INFO_DADO_PL1 = "Dado PL1: "
INFO_DADO_COM = "Dado COM: "

class Dado():
	def __init__(self):
		random.seed()

	def tiraDado(self):
			#Tira il dado, ossia un numero casuale tra 1 e 6 (estremi INCLUSI)
		self.dado = random.randint(1,6)
		#self.dado = 40
		return self.dado

"""Non essendoci un widget/view Button in pygame ho dovuto "crearlo".
	Questo Button non è altro che un rettangolo
	con il testo centrato all'interno
"""
class Button():
	def __init__(self, crazyGoose, game, x, y, width, height, text):
		self.crazyGoose = crazyGoose
		self.game = game
		self.x = x
		self.y = y
		self.width = width
		self.height = height
		
		#Crea un oggetto Rect (non lo disegna ancora)
		self.rect = pygame.Rect(x, y, width, height)
		#Crea il rettangolo proprio.
		#					Surface			colore			 Rect
		pygame.draw.rect(self.game.display, self.game.BLACK, self.rect
						 	#se = 0 riempe anche l'interno, se > 0 FA SOLO IL CONTORNO, se < 0 NON DISEGNA
						 , 1
						 	#border-radius (per arrotondare gli angoli)
						 , 4)
		draw_text(self.game, text, 14, self.game.BLACK, (x+self.width/2), (y+self.height/2))
		
		
	def controllaPosCursore(self, mouse):
		#Controlla semplicemente che il cursore si trovi
		# all'interno del button.
		if(self.rect.collidepoint(mouse)):
			self.crazyGoose.tiraDado()
			
	
	def detectMouseOver(self, mouse):
		if(self.rect.collidepoint(mouse)):
			#codice per "evidenziare" il rettangolo
			# (per far capire all'utente che ci sta passando sopra)
			pass
	
	"""Dentro al metodo "Rect.collidepoint(mouse)" non c'è niente di che...
		Ci sarà qualcosa del genere per controllare che il cursore si trovi
		all'interno del rettangolo:
			self.x <= mouse_x <= self.x + self.width
			and self.y <= mouse_y <= self.y + self.height"""
	
class CrazyGoose():
	def __init__(self, game):
		self.game = game
		self.giocoPartito = False
		
	# sovrappone
	def blit_screen(self):
		#Va a disegnare. (a partire dall'angolo in alto a sx, coord. x=0 y=0)
		self.game.window.blit(self.game.display, (0, 0))
		pygame.display.update()
	
	def start(self):
		#Questo metodo è richiamato all'interno di un ciclo, quindi ho settato un flag,
		# per INIZIALIZZARE le variabili una volta soltanto.
		#(volendo il metodo può essere usato per RIAVVIARE il gioco
		#	N.B. BISOGNA RISETTARE A FALSE il FLAG)
		if(self.giocoPartito == False):
			
			self.giocoPartito = True
			self.partitaTerminata = False
			
				#Dentro questa lista ci saranno gli oggetti Casella
			self.caselle = list()
				#Crea il percorso (casuale)
			self.percorso = Percorso()
			
			self.valDadoCOM = "0"
			self.valDadoPL1 = "0"

			self.buttonDadoPL1 = None
			self.player = None
			self.com = None
			self.turnoCOM = None
			
			self.disegnaTutto()
			
	
	def mousePremuto(self, mousePosition):
		#Controllo che sia il turno del PL1 (cioè che sia il suo turno OPPURE l'avversario abbia un fermo)
		if(self.player != None and self.player.turniFermo == 0):
			if(self.player.turnoMio == True or (self.player.turnoMio == False and self.com.turniFermo > 0)):
				if(not self.buttonDadoPL1 == None):
					self.buttonDadoPL1.controllaPosCursore(mousePosition)

	def mouseOver(self, mousePosition):
		if (not self.buttonDadoPL1 == None):
			self.buttonDadoPL1.detectMouseOver(mousePosition)
	
	def disegnaTutto(self):
			#Se è il primo giro deve stabilire chi sarà il primo giocatore a giocare
		flagPrimoGiro = (self.player == None and self.com == None)
		
		""" "ripulisce" tutto.
			  Purtroppo pygame funziona in questo modo, non posso "spostare" un elemento
			  posso solo ripartire da foglio bianco a disegnare"""
		self.game.display.fill(self.game.WHITE)
		
			#Riempe la lista di caselle, cioè "disegna" gli ellissi.
		self.posizionaLeCaselle()
			#Scrive nelle caselle il loro effetto
		self.riempiCaselle()
		
		draw_text(self.game, INFO_PL1, 15, self.game.BLACK, 50, 20)
		draw_text(self.game, INFO_DADO_PL1, 14, self.game.BLACK, 55, 60)
		self.scriviEffetti(self.player, True)


		#Il primo argomento è self xke necessita di questa classe per
		# richiamarne il metodo "tiraDado"
		self.buttonDadoPL1 = Button(self, self.game, 100, 43, 35, 35, self.valDadoPL1)
		
		draw_text(self.game, INFO_COM, 14, self.game.BLACK, 930, 20)
		draw_text(self.game, (INFO_DADO_COM+self.valDadoCOM), 14, self.game.BLACK, 917, 60)
		self.scriviEffetti(self.com, False)


		if(self.player == None or self.player.posizione == 0):
			#Se posizione == 0 vuol dire che è nella casella iniziale (quella "prima" del percorso)
			self.player = Giocatore(self, self.game, self.caselle, self.percorso, "PL1")
		else:
			self.player.spostaPedina(self.com.posizione)
			
		if(self.com == None or self.com.posizione == 0):
			#Se posizione == 0 vuol dire che è nella casella iniziale (quella "prima" del percorso)
			self.com = Giocatore(self, self.game, self.caselle, self.percorso, "COM")
		else:
			self.com.spostaPedina(self.player.posizione)
			
		if(flagPrimoGiro):
			# Decide chi incomincia, tira il dado e vede se il numero tirato è pari o dispari
			# (tra 1 e 6 ci sono 3 pari e 3 dispari, perciò 50% possbilità a testa)
			if (Dado().tiraDado() % 2 == 0):	#Dado().tiraDado() % 2 == 0
				self.player.turnoMio = True
				self.com.turnoMio = False
				
				self.segnalaChiTocca(True)
			else:
				self.player.turnoMio = False
				self.com.turnoMio = True
				
				#Fa giocare il COM. Aspetterà 2 sec e lancierà il dado....
				# il tutto in un altro thread (altro processo) xkè sennò
				# si bloccherebbe TUTTO per 2 secondi (a causa della sleep di 2 sec)
				self.turnoCOM = threading.Thread(target=self.toccaAlCOM)
				self.turnoCOM.start()
			
				self.segnalaChiTocca(False)
		
		self.blit_screen()
	

	def scriviEffetti(self, player_com, isPL1):
		if(player_com != None):
			pxTesto = 13
			font = pygame.font.Font(self.game.font_name, pxTesto)
			xRect = 20

			if(player_com.ultimoMsg != ""):
				dimTesto = font.size(player_com.ultimoMsg)
				if(not isPL1):
					xRect = self.game.DISPLAY_W - 20 - dimTesto[0]

				yRect = 90
				#3.5px distanza dal bordo dx e sx del rettangolo
				xText = xRect+(dimTesto[0])/2 +3.5
				#(i "/2" xkè scriverò il testo CENTRATO, quindi mi servono le x e y CENTRALI)
				#1px distanza dal bordo sup. e inf. del rettangolo
				yText = yRect+(dimTesto[1]/2) +1

				#dimTesto[0]+7  = larghezza del testo più distanza dal bordo a dx + sx
				#dimTesto[0]+2  = larghezza del testo più distanza dal bordo in alto + in basso
				pygame.draw.rect(self.game.display, self.game.BLACK, pygame.Rect(xRect, yRect, dimTesto[0]+7, dimTesto[1]+2), 1)
				draw_text(self.game, player_com.ultimoMsg, pxTesto, self.game.BLACK, xText, yText)

				if(player_com.penultimoMsg != ""):
					dimTesto = font.size(player_com.penultimoMsg)
					if(not isPL1):
						xRect = self.game.DISPLAY_W - 20 - dimTesto[0]

					yRect += (dimTesto[1]+2+1)
					#3.5px distanza dal bordo dx e sx del rettangolo
					xText = xRect+(dimTesto[0])/2 +3.5
					#(i "/2" xkè scriverò il testo CENTRATO, quindi mi servono le x e y CENTRALI)
					#1px distanza dal bordo sup. e inf. del rettangolo
					yText = yRect+(dimTesto[1]/2) +1
					
					pygame.draw.rect(self.game.display, self.game.BLACK, pygame.Rect(xRect, yRect, dimTesto[0]+7, dimTesto[1]+2), 1)
				
					draw_text(self.game, player_com.penultimoMsg, pxTesto, self.game.BLACK, xText, yText)
		

	def tiraDado(self):
			#Se la partita è terminata "blocca" la funzionalità del dado
		if(not self.partitaTerminata):
			numEstratto = Dado().tiraDado()
			if(self.player.turnoMio):
				self.avanzaPlayer1(numEstratto)
			else:
				self.turnoCOM = threading.Thread(target=self.toccaAlCOM, args=(numEstratto,))
				self.turnoCOM.start()
			

	def toccaAlCOM(self, numEstratto=-1):
		#Se il num del dado non viene passato viene settato a -1 e allora lancia il dado
		if(numEstratto == -1):
			numEstratto = Dado().tiraDado()
		
		#Non voglio bloccare il PL1 per 2 secondi, quindi se il COM ha un fermo non
		# faccio la sleep di 2 sec (dopo aver decrementato il fermo lancia il metodo
		# per far avanzare PL1)
		if(self.com.turniFermo == 0):
			time.sleep(2)

		self.avanzaCOM(numEstratto)

	def avanzaPlayer1(self, numEstratto):
		#numEstratto = 40
		if(self.player.turniFermo > 0):
			# Se il PL1 ha beccato un fermo in precedenza deve "consumarlo"
			# (sarà come se avesse tirato l'avversario)
			self.player.turniFermo -= 1
			
			self.player.turnoMio = False
			self.com.turnoMio =	 True

			self.segnalaChiTocca(False)

			self.turnoCOM = threading.Thread(target=self.toccaAlCOM, args=(numEstratto,))
			self.turnoCOM.start()
		else:
			#(La prossima volta che verrà richiamato "disegnaTutto" il numero
			# del dado sarà cambiato)
			self.valDadoPL1 = str(numEstratto)

			toccaAncoraA_Me = self.player.avanza(numEstratto)
			if(not self.player.vincitore):
					#avanza() ritorna self.turnoMio, perciò
					# se ritorna True non tocca all'avversario (gli setto False)
					# se ritorna False tocca all'avversario (gli setto True)
				self.com.turnoMio = not toccaAncoraA_Me
					
					#se prende un fermo ANNULLA il fermo dell'avversario 
					#(SENNÒ NESSUNO GIOCHEREBBE PIÙ per alcuni turni)
				if(self.player.turniFermo > 0):
					self.com.turniFermo = 0
						#Ora il PL1 ha un fermo, quindi tocca all'avversario sicuro
					self.segnalaChiTocca(False)
					
					#Lancerà il dado, entrerà in avanzaPlayer1 che
					# decrementerà il suo fermo e lancerà avanzaCOM
					self.tiraDado()
				else:
					if(self.com.turniFermo > 0):
							#Se l'avversario ha un fermo al 100% tocca al PL1...
						self.segnalaChiTocca(True)
					else:
						#L'avversario non ha un fermo MA non è detto che tocchi a lui 
						# (PL1 potrebbe aver preso un TIRA DI NUOVO) quindi controllo
						if(self.com.turnoMio):
							self.segnalaChiTocca(False)

							self.turnoCOM = threading.Thread(target=self.toccaAlCOM)
							self.turnoCOM.start()
						else:
							self.segnalaChiTocca(True)

			else:
					#Segnala PL1 come vincitore e fa TERMINARE la partita
				self.segnalaVincitore(True)
				self.partitaTerminata = True


	def avanzaCOM(self, numEstratto):
		#numEstratto = 40
		#(Per i commenti VEDI "avanzaPlayer1")
		
		if(self.com.turniFermo > 0):
			self.com.turniFermo -= 1

			self.player.turnoMio = True
			self.com.turnoMio = False

			self.segnalaChiTocca(True)

			self.avanzaPlayer1(numEstratto)
		else:
			self.valDadoCOM = str(numEstratto)

			toccaAncoraA_Me = self.com.avanza(numEstratto)
			
			if(not self.com.vincitore):
				self.player.turnoMio = not toccaAncoraA_Me
				if(self.com.turniFermo > 0):
					self.player.turniFermo = 0
					self.segnalaChiTocca(True)
				else:
					if(self.player.turniFermo > 0):
						self.segnalaChiTocca(False)
						self.tiraDado()
					else:
						if(self.player.turnoMio):
							self.segnalaChiTocca(True)
						else:
							self.segnalaChiTocca(False)
							self.turnoCOM = threading.Thread(target=self.toccaAlCOM)
							self.turnoCOM.start()
			else:
				self.segnalaVincitore(False)
				self.partitaTerminata = True


	def segnalaChiTocca(self, turnoPlayer1):
		#TODO ottimizzare, fa la stessa cosa, basta mettere var 1 o -1
		if(turnoPlayer1):
			pygame.draw.rect(self.game.display, self.game.RED, pygame.Rect(15, 10, 70, 20), 1)
			
			#se < 0 NON LO DISEGNA, se = 0 riempe anche l'interno, se > 0 FA SOLO IL CONTORNO
			pygame.draw.rect(self.game.display, self.game.RED, pygame.Rect(867, 10, 125, 20), -1)
		else:	
			#se < 0 NON LO DISEGNA, se = 0 riempe anche l'interno, se > 0 FA SOLO IL CONTORNO
			pygame.draw.rect(self.game.display, self.game.RED, pygame.Rect(15, 10, 70, 20), -1)

			pygame.draw.rect(self.game.display, self.game.RED, pygame.Rect(867, 10, 125, 20), 1)

		self.blit_screen()

	def segnalaVincitore(self, haVintoPlayer1):
		time.sleep(1)
		if(haVintoPlayer1):
			# print("HA VINTO e sono in PLAYER  !!!")
			self.game.gameWin()

		else:
			# print("HA PERSO E sono in COM  !!!")
			self.game.gameOver()
			
	def riempiCaselle(self):
		"""Cicla per ogni casella e controlla, se la posizione della casella
		 c'è tra le chiavi del dizionario del percorso allora "carico" l'effetto
		 (prendendo il codice contenuto nel valore con quella chiave, il codice della casella),
		 altrimenti non metto nulla (uso il codice -1)"""
		i=1
		for i in range(QTA_CASELLE_TOTALI+1):
			if(i in self.percorso.dictCaselle.keys()):
				self.caselle[i-1].settaEffetto(self.percorso.dictCaselle[i], self.game.font_name)
			else:	#chiave inesistente (cioè posizione VUOTA)
				self.caselle[i-1].settaEffetto(-1, self.game.font_name)


	def posizionaLeCaselle(self):
		self.caselle.append( Casella(self.game.display, 160, 600) )
		self.caselle.append( Casella(self.game.display, 270, 600) )
		self.caselle.append( Casella(self.game.display, 380, 600) )
		self.caselle.append( Casella(self.game.display, 490, 600) )
		self.caselle.append( Casella(self.game.display, 600, 600) )
		self.caselle.append( Casella(self.game.display, 710, 600) )
		
		self.caselle.append( Casella(self.game.display, 790, 545) )
		self.caselle.append( Casella(self.game.display, 840, 475) )
		self.caselle.append( Casella(self.game.display, 850, 405) )
		self.caselle.append( Casella(self.game.display, 840, 335) )
		self.caselle.append( Casella(self.game.display, 790, 265) )
		
		self.caselle.append( Casella(self.game.display, 710, 210) )
		self.caselle.append( Casella(self.game.display, 600, 180) )
		self.caselle.append( Casella(self.game.display, 490, 180) )
		self.caselle.append( Casella(self.game.display, 380, 180) )
		self.caselle.append( Casella(self.game.display, 270, 180) )
		self.caselle.append( Casella(self.game.display, 170, 220) )

		self.caselle.append( Casella(self.game.display, 125, 300) )
		self.caselle.append( Casella(self.game.display, 125, 385) )
		self.caselle.append( Casella(self.game.display, 125, 470) )

		self.caselle.append( Casella(self.game.display, 215, 530) )
		self.caselle.append( Casella(self.game.display, 325, 530) )
		self.caselle.append( Casella(self.game.display, 435, 530) )
		self.caselle.append( Casella(self.game.display, 545, 530) )
		self.caselle.append( Casella(self.game.display, 653, 515) )
		
		self.caselle.append( Casella(self.game.display, 735, 465) )
		self.caselle.append( Casella(self.game.display, 735, 380) )
		self.caselle.append( Casella(self.game.display, 680, 305) )
		
		self.caselle.append( Casella(self.game.display, 585, 260) )
		self.caselle.append( Casella(self.game.display, 475, 260) )
		self.caselle.append( Casella(self.game.display, 365, 260) )
		
		self.caselle.append( Casella(self.game.display, 265, 295) )
		self.caselle.append( Casella(self.game.display, 230, 370) )
		self.caselle.append( Casella(self.game.display, 270, 445) )
		
		self.caselle.append( Casella(self.game.display, 380, 460) )
		self.caselle.append( Casella(self.game.display, 490, 460) )
		self.caselle.append( Casella(self.game.display, 590, 440) )
		
		self.caselle.append( Casella(self.game.display, 625, 370) )
		self.caselle.append( Casella(self.game.display, 525, 335) )
		self.caselle.append( Casella(self.game.display, 370, 340, 105, 95) )

File: webApp_server/public/res_static_gioco/js/crazyGoose.js
const INFO_PL1 = "Tu (PL1)"
const INFO_COM = "Computer (COM)"
const INFO_DADO_PL1 = "Dado PL1: "
const INFO_DADO_COM = "Dado COM: "

class Dado {
    tiraDado() {
        //num casuale tra 0 e 5 => +1 => num casuale tra 1 e 6
        return (Math.floor(Math.random() * 6) + 1)
    }
}

class CrazyGoose {
    constructor() {
        this.giocoPartito = false
    }

    start() {
        //Questo metodo è richiamato all'interno di un ciclo, quindi ho settato un flag,
        // per INIZIALIZZARE le variabili una volta soltanto.
        //(volendo il metodo può essere usato per RIAVVIARE il gioco
        //	N.B. BISOGNA RISETTARE A false il FLAG)
        if (this.giocoPartito == false) {
            this.giocoPartito = true
            this.partitaTerminata = false

            //Dentro questa lista ci saranno gli oggetti Casella
            this.caselle = []
                //Crea il percorso (casuale)
            this.percorso = new Percorso()

            this.buttonDadoPL1 = null
            this.player = null
            this.com = null
            this.turnoCOM = null
            this.timeoutCOM = null

            this.disegnaTutto()
        }
    }

    disegnaTutto(giocDaNonDisegnare = null) {
        //Riempe la lista di caselle, cioè "disegna" gli ellissi.
        this.posizionaLeCaselle()
            //Scrive nelle caselle il loro effetto
        this.riempiCaselle()

        //this.scriviEffetti(this.player, true)

        //Il primo argomento è this xke necessita di questa classe per
        // richiamarne il metodo "tiraDado"
        this.buttonDadoPL1 = document.getElementById("dado_pl1")
        this.buttonDadoPL1.addEventListener("click", () => {

            //Controllo che sia il turno del PL1 (cioè che sia il suo turno o che l'avversario abbia un fermo)
            if (this.player != null && this.player.turniFermo == 0) {
                if (this.player.turnoMio == true || (this.player.turnoMio == false && this.com.turniFermo > 0)) {
                    this.tiraDado()
                }
            }

        })

        this.player = new Giocatore(this, this.caselle, this.percorso, "PL1")
        this.com = new Giocatore(this, this.caselle, this.percorso, "COM")


        // Decide chi incomincia, tira il dado e vede se il numero tirato è pari o dispari
        // (tra 1 e 6 ci sono 3 pari e 3 dispari, perciò 50% possbilità a testa)
        let random = (new Dado()).tiraDado()
        if (random % 2 == 0) {
            this.player.turnoMio = true
            this.com.turnoMio = false

            this.segnalaChiTocca(true)
        } else {
            this.player.turnoMio = false
            this.com.turnoMio = true

            this.toccaAlCOM()

            this.segnalaChiTocca(false)
        }
    }

    scriviEffetti(player_com, isPL1) {
        let eff1 = null
        let eff2 = null
        if (isPL1) {
            eff1 = document.getElementById("eff1_pl1")
            eff2 = document.getElementById("eff2_pl1")
        } else {
            eff1 = document.getElementById("eff1_com")
            eff2 = document.getElementById("eff2_com")
        }
        eff1.style.display = "block"
        if (player_com.penultimoEff != "") {
            eff2.style.display = "block"
        }

        eff1.innerHTML = player_com.ultimoEff
        eff2.innerHTML = player_com.penultimoEff
    }

    stopCOM() {
        if (this.timeoutCOM != null) {
            clearTimeout(this.timeoutCOM)
            clearTimeout(this.timeoutCOM)
        }
    }

    tiraDado() {
        //Se la partita è terminata "blocca" la funzionalità del dado (anche se in realtà non serve xke non dovrebbe più vederlo il dado, e quindi cliccarlo)
        if (!this.partitaTerminata) {
            let numEstratto = (new Dado()).tiraDado()

            if (this.player.turnoMio) {
                this.avanzaPlayer1(numEstratto)
            } else {
                this.toccaAlCOM()
            }
        }
    }

    toccaAlCOM(numEstratto = -1) {
        //Se il num del dado non viene passato (è settato a -1) lancia il dado
        if (numEstratto == -1) {
            numEstratto = (new Dado()).tiraDado()
        }

        //Non devo bloccare il PL1 per 1 secondo, quindi se il COM ha un fermo non
        // faccio la sleep di 1 sec (dopo aver decrementato il fermo lancia il metodo
        // per far avanzare PL1 quindi tecnicamente si muove il PL1 e si deve muovere
        // subito non dopo 1 sec)
        if (this.com.turniFermo == 0) {
            //(* * * setTimeout esegue, in una sorta di altro processo (quindi non blocca il codice),
            // la funzione passata dopo tot millisecondi * * *)
            this.timeoutCOM = setTimeout(() => { this.avanzaCOM(numEstratto) }, 1000)
        } else {
            this.avanzaCOM(numEstratto)
        }
    }

    avanzaPlayer1(numEstratto) {
        if (this.player.turniFermo > 0) {
            // Se il PL1 ha beccato un fermo in precedenza deve "consumarlo"
            // (sarà come se avesse tirato l'avversario)
            this.player.turniFermo -= 1

            this.player.turnoMio = false
            this.com.turnoMio = true

            this.segnalaChiTocca(false)

            this.toccaAlCOM(numEstratto)
        } else {
            /*(La prossima volta che verrà richiamato "disegnaTutto" il numero
                del dado sarà cambiato) */
            this.buttonDadoPL1.innerHTML = numEstratto

            this.player.avanza(numEstratto)

            //(* * * setinterval esegue, in una sorta di altro processo (quindi non blocca il codice),
            // la funzione passata OGNI tot millisecondi (finchè non lo si ferma con clearTimeout
            // continuerà ad eseguire ogni tot ms la funzione data) * * *)
            let idIntervalFineAvanza = setInterval(() => {
                if (this.player.isMoving == false) {
                    clearInterval(idIntervalFineAvanza)

                    if (!this.player.vincitore) {
                        this.com.turnoMio = !this.player.turnoMio

                        //se prende un fermo ANNULLA il fermo dell'avversario 
                        //(SENNÒ NESSUNO GIOCHEREBBE PIÙ per alcuni turni)
                        if (this.player.turniFermo > 0) {
                            this.com.turniFermo = 0
                                //Ora il PL1 ha un fermo, quindi tocca all'avversario sicuro
                            this.segnalaChiTocca(false)

                            //Lancerà il dado, entrerà in avanzaPlayer1 che
                            // decrementerà il suo fermo e lancerà avanzaCOM
                            this.tiraDado()
                        } else {
                            if (this.com.turniFermo > 0) {
                                //Se l'avversario ha un fermo al 100% tocca al PL1...
                                this.segnalaChiTocca(true)
                            } else {
                                //L'avversario non ha un fermo MA non è detto che tocchi a lui 
                                // (PL1 potrebbe aver preso un TIRA DI NUOVO) quindi controllo
                                if (this.com.turnoMio) {
                                    this.segnalaChiTocca(false)

                                    this.toccaAlCOM()
                                } else {
                                    this.segnalaChiTocca(true)
                                }
                            }
                        }
                    } else {
                        //Segnala PL1 come vincitore e fa TERMINARE la partita
                        this.segnalaVincitore(true)
                        this.partitaTerminata = true
                    }

                }
            }, 100)
        }
    }

    avanzaCOM(numEstratto) {
        //* * *  Per i commenti VEDI "avanzaPlayer1" * * *

        if (this.com.turniFermo > 0) {
            this.com.turniFermo -= 1

            this.player.turnoMio = true
            this.com.turnoMio = false

            this.segnalaChiTocca(true)

            this.avanzaPlayer1(numEstratto)
        } else {
            document.getElementById("dado_com").innerHTML = numEstratto

            this.com.avanza(numEstratto)

            let idIntervalFineAvanza = setInterval(() => {
                if (this.com.isMoving == false) {
                    clearInterval(idIntervalFineAvanza)

                    if (!this.com.vincitore) {
                        this.player.turnoMio = !this.com.turnoMio
                        if (this.com.turniFermo > 0) {
                            this.player.turniFermo = 0
                            this.segnalaChiTocca(true)
                        } else {
                            if (this.player.turniFermo > 0) {
                                this.segnalaChiTocca(false)
                                this.tiraDado()
                            } else {
                                if (this.player.turnoMio) {
                                    this.segnalaChiTocca(true)
                                } else {
                                    this.segnalaChiTocca(false)
                                    this.toccaAlCOM()
                                }
                            }
                        }
                    } else {
                        this.segnalaVincitore(false)
                        this.partitaTerminata = true
                    }
                }
            }, 100)
        }
    }

    segnalaVincitore(pl1_ha_vinto) {
        if (pl1_ha_vinto) {
            alert("ANCORA DA IMPLEMENTARE !!!!          PL1 HA VINTO.")
        } else {
            alert("ANCORA DA IMPLEMENTARE !!!!          COM HA VINTO.")
        }

    }


    segnalaChiTocca(tocca_al_pl1) {
        let chiTocca = null
        let chiAspetta = null
        if (tocca_al_pl1) {
            chiTocca = document.getElementById("info_pl1")
            chiAspetta = document.getElementById("info_com")
        } else {
            chiTocca = document.getElementById("info_com")
            chiAspetta = document.getElementById("info_pl1")
        }
        chiTocca.style.padding = "1px"
        chiTocca.style.border = "2px solid red"

        //rimuovo il bordo (lo nascondo in pratica)
        chiAspetta.style.padding = "0px"
        chiAspetta.style.border = "2px none red"
    }

    riempiCaselle() {
        /*Cicla per ogni casella e controlla, se la posizione della casella
         c'è tra le chiavi del dizionario del percorso allora "carico" l'effetto
         (prendendo il codice contenuto nel valore con quella chiave, il codice della casella),
         altrimenti non metto nulla (uso il codice -1)*/
        //TODO controllare bene se gira per tutte le caselle
        for (let i = 1; i <= 40; i++) {
            let casella = this.caselle[i - 1]
            try {
                casella.settaEffetto(this.percorso.dictCaselle[i])
            } catch (err) {
                //chiave inesistente (cioè posizione VUOTA)
                casella.settaEffetto(-1)
            }
        }
    }

    posizionaLeCaselle() {
        this.caselle.push(new Casella(1, 130, 525))
        this.caselle.push(new Casella(2, 240, 525))
        this.caselle.push(new Casella(3, 350, 525))
        this.caselle.push(new Casella(4, 460, 525))
        this.caselle.push(new Casella(5, 570, 525))
        this.caselle.push(new Casella(6, 680, 525))

        this.caselle.push(new Casella(7, 785, 470))
        this.caselle.push(new Casella(8, 835, 400))
        this.caselle.push(new Casella(9, 845, 330))
        this.caselle.push(new Casella(10, 835, 260))
        this.caselle.push(new Casella(11, 785, 190))

        this.caselle.push(new Casella(12, 710, 135))
        this.caselle.push(new Casella(13, 600, 105))
        this.caselle.push(new Casella(14, 490, 105))
        this.caselle.push(new Casella(15, 380, 105))
        this.caselle.push(new Casella(16, 270, 105))
        this.caselle.push(new Casella(17, 170, 145))

        this.caselle.push(new Casella(18, 125, 225))
        this.caselle.push(new Casella(19, 125, 310))
        this.caselle.push(new Casella(20, 125, 395))

        this.caselle.push(new Casella(21, 215, 455))
        this.caselle.push(new Casella(22, 325, 455))
        this.caselle.push(new Casella(23, 435, 455))
        this.caselle.push(new Casella(24, 545, 455))
        this.caselle.push(new Casella(25, 653, 445))

        this.caselle.push(new Casella(26, 735, 390))
        this.caselle.push(new Casella(27, 735, 305))
        this.caselle.push(new Casella(28, 680, 230))

        this.caselle.push(new Casella(29, 585, 185))
        this.caselle.push(new Casella(30, 475, 185))
        this.caselle.push(new Casella(31, 365, 185))

        this.caselle.push(new Casella(32, 265, 220))
        this.caselle.push(new Casella(33, 230, 295))
        this.caselle.push(new Casella(34, 270, 370))

        this.caselle.push(new Casella(35, 380, 385))
        this.caselle.push(new Casella(36, 490, 385))
        this.caselle.push(new Casella(37, 590, 365))

        this.caselle.push(new Casella(38, 625, 295))
        this.caselle.push(new Casella(39, 525, 260))
        this.caselle.push(new Casella(40, 375, 280, true))
    }
}